src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -876,6 +876,517 @@ uct_ib_mlx5_devx_dereg_keys(uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mem_t *memh)
+    }
+
+    /* Add UMR mkey to mkey pool for reuse & cleanup */
+    ucs_list_add_head(&md->umr_mkey_pool, &umr->super);
+    ucs_trace("%s: put " UCT_IB_MLX5_UMR_FMT " into pool (%ld)",
+              uct_ib_device_name(&md->super.dev), UCT_IB_MLX5_UMR_ARG(umr),
+              ucs_list_length(&md->umr_mkey_pool));
+
+    return UCS_OK;
+
+destroy_mkey:
+    uct_ib_mlx5_devx_umr_mkey_destroy(md, umr);
+    return UCS_ERR_IO_ERROR;
+}
+
+static uct_ib_mlx5_devx_umr_t *
+uct_ib_mlx5_devx_umr_mkey_pool_get(uct_ib_mlx5_md_t *md)
+{
+    uct_ib_mlx5_devx_umr_t *umr;
+
+    if (ucs_list_is_empty(&md->umr_mkey_pool)) {
+        umr = uct_ib_mlx5_devx_umr_mkey_create(md);
+    } else {
+        umr = ucs_list_extract_head(&md->umr_mkey_pool, uct_ib_mlx5_devx_umr_t,
+                                    super);
+
+        ucs_trace("%s: extracted from pool " UCT_IB_MLX5_UMR_FMT,
+                  uct_ib_device_name(&md->super.dev), UCT_IB_MLX5_UMR_ARG(umr));
+    }
+
+    return umr;
+}
+
+static ucs_status_t
+uct_ib_mlx5_devx_umr_mkey_bind(uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mr_t *mr,
+                               uct_ib_mlx5_devx_umr_t **umr)
+{
+    uct_ib_device_t *ibdev = &md->super.dev;
+    struct ibv_mr *ib_mr   = mr->super.ib;
+    struct ibv_wc wc       = {};
+    struct ibv_mw mw       = {
+        .rkey = 0,
+        .type = IBV_MW_TYPE_1
+    };
+    struct ibv_send_wr wr  = {
+        .wr_id      = 0,
+        .next       = NULL,
+        .opcode     = IBV_WR_BIND_MW,
+        .send_flags = IBV_SEND_INLINE | IBV_SEND_SIGNALED,
+        .bind_mw    = {
+            .mw        = &mw,
+            .rkey      = 0,
+            .bind_info = {
+                .mr              = ib_mr,
+                .length          = ib_mr->length,
+                .addr            = (uint64_t)ib_mr->addr,
+                .mw_access_flags = UCT_IB_MLX5_MD_UMEM_ACCESS
+            }
+        }
+    };
+    int ret;
+    struct ibv_send_wr *bad_wr;
+
+    *umr = uct_ib_mlx5_devx_umr_mkey_pool_get(md);
+    if (NULL == *umr) {
+        goto err;
+    }
+
+    /* Set UMR lkey to be registered with given MR */
+    mw.rkey         = (*umr)->umr_mkey->lkey;
+    wr.bind_mw.rkey = (*umr)->umr_mkey->lkey;
+
+    if (ibv_post_send(md->umr_qp, &wr, &bad_wr) != UCS_OK) {
+        ucs_error("%s: failed to post UMR mkey registration WR for addr %p len "
+                  "%zu lkey 0x%x: %m", uct_ib_device_name(ibdev),
+                  ib_mr->addr, ib_mr->length, ib_mr->lkey);
+
+        goto invalidate;
+    }
+
+    /* Poll for completion */
+    while ((ret = ibv_poll_cq(md->umr_cq, 1, &wc)) == 0);
+
+    if ((ret < 0) || (wc.status != IBV_WC_SUCCESS)) {
+        ucs_error("%s: failed to poll CQ for UMR mkey registration for addr %p "
+                  " len %zu lkey 0x%x ret %d status %d",
+                  uct_ib_device_name(ibdev), ib_mr->addr, ib_mr->length,
+                  ib_mr->lkey, ret, wc.status);
+
+        goto invalidate;
+    }
```

Comment:
the sme sequence is in `uct_ib_mlx5_devx_umr_mkey_invalidate`, maybe a separate inline function is worth creating?

Reply:
Thanks, totally makes sense, added